### DISCUSSION TITLE
[Mobile] Critical Issue / SDK Unusable - Inconsistent wallets' interactions

### DISCUSSION THREAD

**Discussion by nikolay-and** [WalletConnect/WalletConnectFlutterV2#287] [Q&A]
We are developing a dapp using flutter. Last year we integrated WalletConnect v2 and it worked flawlessly for the most part. However, as of the previous nearly 2 months we have a major issue seriously impacting our application's behaviour. The issue can be summarised with the following steps:

1. The user authenticatiates with their wallet and establishes a session with wallet connect (Either using an iOS or Android device)
2. From the dapp the user initiate a transaction with `eth_sendTransaction` or any other supported method by the selected chain
3. The connected wallet is being open and the prompt confirming the transaction appears as expected
4. The user confirms it and everything is processed accordingly
5. The user waits for around 1-2 minutes
6. The user tries to perform a new transaction (the same or another one)
7. The connected wallet is being open and no prompt is being shown regardless of how much time the user waits
8. Every other attempt to perform a transaction regardless of its method results in the wallet never showing the prompt again

The only solution to tackle the issues that we have found is to manually disconnect the session and to kill the wallet application, which of course is not an appropriate ux solution. Initially we thought that it was a Metamask problem, as we mostly tested with it, but after testing this scenario on Trust and Rainbow, where behaviour was exactly the same we concluded that it should have been a wallet connect issues. we've tried reverting the `walletconnect_flutter_v2 ` library to a significantly older version (2.2.3 -> 2.1.5) but of no avail. We've used Avalanche and Etehreum and regardless of the chain the problem was still consistent. Recently we've migrated to `web3modal_flutter` (Version 3.1.2) but the problems persists with this sdk as well. Additionally the problem was reproducible with the sample app provided by wallet connect's documentation for Flutter. In addition we created a sample iOS app using the `web3modal-swift` library and unfortunately the problem was reproducible by applying the very same steps. 

We have also noticed that when using the Zerion application and when having metamask linked performing a swap operation, even though it worked as expected immediately after establishing a connection, after 1-2 minutes no more prompts ever appeared again in metamask until the session was manually disconnected and metamask was killed, exactly the same issues that we have in our project.

This is a major issues for us as we expect our customers to have a multitude transactions to be performed, otherwise our business logic is substantially impacted and we won't be able to release our mobile dapps until the issue is resolved

Below I am attaching a video, in which the issue is demonstrated using the flutter's sample app provided by wallet connect with the button `Send2` performing a simple `personal_sign`
[Video Link](https://drive.google.com/file/d/1IzxBOV7Kqkf-moc5C20-xYi3B2c3HI9B/view?usp=share_link)

[Github Repo Sample App](https://github.com/nikolay-and/web3modal_flutter_sample) That's the sample app, based on which is the recorded video.

**Comment by randomrandom** [comment]
I have the same issue and it's horrible. Saw this reported issue here: https://github.com/orgs/WalletConnect/discussions/4358

maybe @glitch-txs has a better idea why it happens

**Comment by NasirMaDev** [comment]
> The issue can be summarised with the following steps:
> 
> 1. The user authenticatiates with their wallet and establishes a session with wallet connect (Either using an iOS or Android device)
> 2. From the dapp the user initiate a transaction with `eth_sendTransaction` or any other supported method by the selected chain
> 3. The connected wallet is being open and the prompt confirming the transaction appears as expected
> 4. The user confirms it and everything is processed accordingly
> 5. The user waits for around 1-2 minutes
> 6. The user tries to perform a new transaction (the same or another one)
> 7. The connected wallet is being open and no prompt is being shown regardless of how much time the user waits
> 8. Every other attempt to perform a transaction regardless of its method results in the wallet never showing the prompt again

Working on native iOS app, exactly same issue and session handlers doesnt show any error or warnings too.

**Reply by randomrandom** [reply]
we also tested on native iOS, I can also confirm that we hit the same error

**Reply by NasirMaDev** [reply]
I am using iOS 16.7.5 currently but also tested on iOS 17 before.
Regarding wait time, it doesn't seem to matter because it takes my app one to two minutes to select the assets, amount, recipient address, and priority before signing a new transaction.

**Comment by glitch-txs** [comment]
cc @quetool

**Reply by quetool** [reply]
Hello @nikolay-and please allow me some time to investigate this with the rest of the team. Thanks.

**Reply by randomrandom** [reply]
Hi @quetool did you manage to reproduce? We are also facing this issue and it's making the wallet connect approach practically unusable. No user will end the session, kill the app and reconnect the wallet after each transaction

**Comment by quetool** [comment]
Just to understand:

- You are saying that this behavior is happening both with flutter SDK and swift SDK
- You are saying this behavior is reproducible with Metamask, Trust and Rainbow
- Did you try using a different nonce on every new transaction?
- Do you have latest versions of wallets?
- Does this happens equally in Android and iOS versions of the respective wallets?

A couple of things after trying @nikolay-and code:

1. If I try "Send 2" button I get the following error
```
Pairing _onMessageEvent, Received data: {id: 1714676414348783, jsonrpc: 2.0, error: {code: 5000, message: {"code":-32602,"message":"Invalid parameters: must provide a permitted Ethereum address."}}}
[ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: JsonRpcError(code: 5000, message: {"code":-32602,"message":"Invalid parameters: must provide a permitted Ethereum address."})
```
The address seems to be not valid, if I change to the one that is provided by the service everything works as expected.

2. I tried Send 1 button on Sepolia networks and I succeeded every time I've sent SEPs, see transactions https://sepolia.etherscan.io/address/0x59e2f66c0e96803206b6486cdb39029abae834c0

**Reply by nikolay-and** [reply]
`You are saying that this behavior is happening both with flutter SDK and swift SDK`
`You are saying this behavior is reproducible with Metamask, Trust and Rainbow`
`Do you have latest versions of wallets?`
`Does this happens equally in Android and iOS versions of the respective wallets?`

That's correct for all of the items above.

`Did you try using a different nonce on every new transaction?`

I've tried using incremental nonce for every new transaction but without success

As for the sample that I've provided there is no `Send 2` action in my sample code it was only for the video. The equivalent action in the sample project is `Send` and it's a simple personal sign:

```dart
_w3mService.launchConnectedWallet();
await _w3mService.request(
                    topic: _w3mService.session?.topic ?? '',
                    chainId: 'eip155:1', // Connected chain id
                    request: SessionRequestParams(
                      method: 'personal_sign',
                      params: ['Sign this', <Input your account address>],
                    ),
);
```

**Reply by quetool** [reply]
Any chance you want to share your projectId? It is completely safe to share projectId and since I'm struggling to reproduce the behavior maybe looking at your project config would help.

**Reply by nikolay-and** [reply]
I've just pushed the changes in the main branch of the sample project, that I have listed, I've added a project ID of a new project that I configured in wallet connect for testing purposes and if you check out the branch it should be plug and play, no addresses should be entered. I've also confirmed that with the new changes I am able to connect and to reproduce the problem. (Tested with iOS Metamask 7.21.0)

I have also noticed something else. While going to the wallet connect's dashboard from my pc, I've selected to authenticate with metamask. The QR code appeared, I opened metamask on my iPhone and after scanning the QR code metamask immediately prompted me to connect. After establishing a connection it was written on the web app that I have to do a sign from my device in order to login. However, no prompt appeared and I've retried the process a couple of times. After manually disconnecting the wallet connect session from metamask and killing the app and upon retrying the process after a successful QR code based connection the prompt appeared and I was able to sign the transaction, which allowed me to login into the dashboard. This behaviour appeared to be exactly the same as the one I had previously described:
`The only solution to tackle the issues that we have found is to manually disconnect the session and to kill the wallet application`

**Reply by quetool** [reply]
Hello @nikolay-and, I have tested your code, with our own sample wallet and Metamask 7.21.0 as well, I also tested WalletConnect dashboard (cloud.walletconnect.com) with Metamask. Everything worked as expected every time. 

https://github.com/WalletConnect/WalletConnectFlutterV2/assets/14978705/e190cd36-d39d-42ec-81a8-a8bf045bb0ca

https://github.com/WalletConnect/WalletConnectFlutterV2/assets/14978705/a47b573b-1fd0-4f3e-b500-3ad8e365dd52

I honestly don't know how to reproduce your behavior (if possible) so I would need more information if possible

**Reply by randomrandom** [reply]
Hi @quetool wr are experiencing the same issues as the ones @nikolay-and mentioned.. we have noticed recently a more stable behaviour, however we noticed that once you connect the session everything works for X amount of time (multiple transactions can be signed) but after some time and probably killing of the apps - the connection breaks down again. Usually it happens a few hours later or even a day later. Not sure what changed, but it does seem to be a bit better now, still fails after a bit of usage though and users are confused why they don't see any prompts 

The behaviour appears on mobile when doing transactions with Metamask (latest version- 7.21) on the avalanche network

**Reply by randomrandom** [reply]
@quetool for me what seems to reproduce the issue consistently is the following:
1. Connect a wallet
2. Execute a "sign", not send transaction but a sign transaction  that we use for account abstraction use case
3. Complete the tx flow
4. Kill Metamask
5. Session needs to be deleted next time I try to do the same transaction 

It seems to work okay if I kill our app but not Metamask, but if I kill Metamask - the session goes bad

**Reply by quetool** [reply]
Hello @randomrandom! Thanks for the detailed steps! Is this happening only with Metamask, mostly with Metamask or equally across wallets? 

From my POV this seems to be a Metamask related issue, they may be struggling to connect to the relay when cold starting. The improvement you are seeing recently could be indeed because Metamaks released a version with a lot of fixes. It indeed works way better than before but maybe they still have some issues. That being said I will anyway trying to debug/reproduce this behavior in case could be something on our side.

Please keep your feedback coming in case you gain more debug info!

**Reply by randomrandom** [reply]
@quetool tested with rainbow, but seems even worse. We will try a few more

**Reply by quetool** [reply]
Zerion, Trust or OKX would be nice to try. Thanks!

**Reply by randomrandom** [reply]
> Zerion, Trust or OKX would be nice to try. Thanks!

@quetool we tried with 3 wallets (Rainbow, Safe, Zerion), all of them have issue if we do a Sign and then kill the wallet, however if we do send and then kill the wallet it works fine. The seems to be buried in the cases where we do a sign and then kill the wallet app.

Trust we can't try, as it doesn't support avalanche. 

I don't think this is specific only to a specific wallet, either there is a problem with how the signing works with all wallets or a problem which is related to the session management on the WC side

**Reply by randomrandom** [reply]
@quetool we found something new, when using eth_signTypedData_v4 (previously we used eth_signTypedData) - we get the following error with Metamask, after killing the Metamask app (first time it works, after we kill it and try to sign again we get this):
`{\"code\":-32602,\"message\":\"Invalid parameters: must provide an Ethereum address.\"}`

Looks like in Zerion it works though! Let us check again though...

**Reply by randomrandom** [reply]
In Zerion it works consistently

**Reply by quetool** [reply]
Hello @randomrandom! Sorry for the late response. I lost track of this but it seems that it's something on wallet side as on Zerion works fine? However I would like to try to reproduce, please can you provide consistent steps to repro?
@nikolay-and do you have any update?
